refactor(core): add initial implementation of function to replace metadata at runtime

crisbeto opened the pull request:
Adds the new `ɵɵreplaceMedata` function that can be used to replace the metadata of a component class and re-render all instances in place without refreshing the page. The function isn't used anywhere at the moment, but it will be necessary for future functionality.

crisbeto commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -211,6 +211,9 @@ export function compileComponent(type: Type<any>, metadata: Component): void {
       }
       return ngComponentDef;
     },
+    set: (def: ComponentDef<unknown> | null) => {
```
Point for discussion: do we want to have an assertion here that the setter only gets invoked as a result of `replaceMetadata`? I didn't, because:
1. This logic only applies to JIT mode, the property is already writable in AOT-compiled components.
2. There isn't much someone can do by replacing the definition, aside from breaking their own app.

AndrewKushnir replied:
I'm leaning towards keeping the current state and disallow writes to `ngComponentDef`, but we can also do that as a followup (leaving a TODO here for now).

crisbeto replied:
I don't see what we're guarding against here though. The field is already prefixed with `ɵ` which indicates that it's private. On top of that, the field has always been writable for AOT-compiled components so we're only really changing the behavior for JIT mode in production.

AndrewKushnir replied:
JIT mode is still used in TestBed APIs (e.g. when you call `TestBed.overrideTemplate`), but as you've mentioned, it's prefixed by the `ɵ` symbol already, so it's clear that this is a private API. Not feeling strongly about it, we can keep the code as is (as implemented in this PR).

crisbeto commented on packages/core/src/render3/hmr.ts:
```diff
@@ -0,0 +1,220 @@
+ * @param applyMetadata Callback that will apply a new set of metadata on the `type` when invoked.
+ * @codeGenApi
+ */
+export function ɵɵreplaceMedata(type: Type<unknown>, applyMetadata: () => void) {
```
The reason why `applyMetadata` is a callback, rather than passing the new metadata directly, is that the metadata can be more than just the new definition. E.g. we have `setClassMetadata` and `setClassDebugInfo` calls. There can also functions for embedded views. The function allows us to have a closure around all of them to keep them isolated.

crisbeto replied:
I'll add a comment inline.

AndrewKushnir commented on packages/core/src/render3/hmr.ts:
```diff
@@ -0,0 +1,220 @@
+    const trackedViews = getTrackedLViews().values();
+    for (const root of trackedViews) {
+      if (root[FLAGS] & LViewFlags.IsRoot) {
+        recreateMatchingLViews(oldDef, root);
```
@crisbeto I might be Wrong, but I'm wondering if we are doing extra (unnecessary) work:

- we go over all "active" LViews
- for each LView, we recursively traverse all of its children and re-create matching LViews

Is it possible that we'll be re-creating an LView multiple times?

crisbeto replied:
We aren't traversing the descendants of all the LViews though, we're only traversing descendants of root views. My assumption here is that `LViewFlags.IsRoot` indicates that it's the root view of an application. Also worth noting that the `TRACKED_LVIEWS` doesn't actually track every single LView, but only the ones connected to patched data.

crisbeto replied:
Hmm, I see. In that case would it make sense to also check `LView[PARENT] === null`?

crisbeto replied:
Pushed a change to check for `PARENT`. I think that should do it since we assign the parent when a dynamically-created component is attached.

AndrewKushnir replied:
The `LView[PARENT] === null` check makes sense 👍

splincode commented:
@crisbeto Do we expect in the future potential creation of dynamic/runtime component?

crisbeto commented:
It's already possible to create components dynamically, if they have been compiled already. If you mean compiling a component from a string, then no, that's not what this PR is about.

splincode commented:
@crisbeto oh, sorry, thank you for explain

crisbeto committed:
refactor(core): add initial implementation of function to replace metadata at runtime

Adds the new `ɵɵreplaceMedata` function that can be used to replace the metadata of a component class and re-render all instances in place without refreshing the page. The function isn't used anywhere at the moment, but it will be necessary for future functionality.

alxhub commented:
This PR was merged into the repository by commit c44f08748241976d177c432a5ae3ac66a8bfd81c.

The changes were merged into the following branches: main
